modules/library/main/src/main/java/org/geotools/feature/AttributeBuilder.java

```diff
@@ -214,15 +215,15 @@ public void setCRS(CoordinateReferenceSystem crs) {
 
     /** @return The coordinate reference system of the feature, or null if not set. */
     public CoordinateReferenceSystem getCRS(Object geom) {
-        if (crs != null) {
-            return crs;
-        } else if (geom != null && geom instanceof Geometry) {
-            Object userData = ((Geometry) geom).getUserData();
-            if (userData != null && userData instanceof CoordinateReferenceSystem) {
-                return (CoordinateReferenceSystem) userData;
+        if (geom != null && geom instanceof Geometry) {
+            // the geometry in the geometry itself is preferred
```

Comment:
"the geometry in the geometry itself"

Something seems to have gone wrong here :-D
You probably wanted to say, the "CRS in the geometry itself".

The statement "to support multiple geometries/CRS" also confuses me... this one builds a single attribute, and the PR is trying to allow different CRSs in different attributes of the same feature... not different CRSs inside the same column, or not? From the ticket "multiple geometries in features, particularly if they have different CRS."

Ah, another thing.. this is a GeoTools change, please have a GeoTools ticket describing it, and refer to it from the commit message. Not all use GeoTools through GeoServer, a ticket in the GeoTools tracker ends up in the release notes and informs the direct GeoTools users too.

Reply:
It seems that most simple feature vector datastore implementations (at least all of the JDBC, as well as property files) encode the CRS from the data directly into the geometry. This mechanism supports different columns with each their own CRS.

 App-schema however assumes a single CRS and this is the place where it is written into each geometry field of the final type/descriptor structure that is served to the encoder. Previously, the CRS inside the geometry was only used as a fallback in case the single CRS attribute is empty  (which is in practice, never) . The order of preference is changed here so that the CRS attribute becomes the fall-back mechanism (in case there was no CRS inside the data), and in this way, the different CRS in the different geometry columns are passed along into the complex type.